src/pkg/packager/network.go

```diff
@@ -272,3 +285,87 @@ func pullLayer(dst *utils.OrasRemote, desc ocispec.Descriptor, out string) error
+	ref, err := registry.ParseReference(strings.TrimPrefix(url, utils.OCIURLPrefix))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse OCI reference: %w", err)
+	}
+	src, err := utils.NewOrasRemote(ref)
+	if err != nil {
+		return nil, err
+	}
+	manifest, err := getManifest(src)
+	if err != nil {
+		return nil, err
+	}
+	zarfYamlDescriptor := utils.Find(manifest.Layers, func(layer ocispec.Descriptor) bool {
+		return layer.Annotations[ocispec.AnnotationTitle] == config.ZarfYAML
+	})
+	zarfYamlBytes, err := content.FetchAll(src.Context, src, zarfYamlDescriptor)
+	if err != nil {
+		return nil, err
+	}
+	pkg := types.ZarfPackage{}
+	err = goyaml.Unmarshal(zarfYamlBytes, &pkg)
+	if err != nil {
+		return nil, err
+	}
+	images := []string{}
+	for _, name := range requestedComponents {
+		component := utils.Find(pkg.Components, func(component types.ZarfComponent) bool {
+			return component.Name == name
+		})
+		if component.Name == "" {
+			return nil, fmt.Errorf("component %s does not exist in this package", name)
+		}
+	}
+	for _, component := range pkg.Components {
+		// If we requested this component, or it is required, we need to pull its images
+		if utils.SliceContains(requestedComponents, component.Name) || component.Required {
+			images = append(images, component.Images...)
+		}
+		// If we did not request this component, but it is required, we need to pull it
+		if !utils.SliceContains(requestedComponents, component.Name) && component.Required {
+			layers = append(layers, filepath.Join(config.ZarfComponentsDir, component.Name))
+		}
+	}
+
+	if len(images) > 0 {
+		indexDescriptor := utils.Find(manifest.Layers, func(layer ocispec.Descriptor) bool {
+			return layer.Annotations[ocispec.AnnotationTitle] == indexPath
+		})
+		indexBytes, err := content.FetchAll(src.Context, src, indexDescriptor)
+		if err != nil {
+			return nil, err
+		}
+		index := ocispec.Index{}
+		err = json.Unmarshal(indexBytes, &index)
+		if err != nil {
+			return nil, err
+		}
+		for _, image := range images {
+			manifestDescriptor := utils.Find(index.Manifests, func(layer ocispec.Descriptor) bool {
+				return layer.Annotations[ocispec.AnnotationBaseImageName] == image
+			})
+			manifestBytes, err := content.FetchAll(src.Context, src, manifestDescriptor)
+			if err != nil {
+				return nil, err
+			}
+			manifest := ocispec.Manifest{}
+			err = json.Unmarshal(manifestBytes, &manifest)
+			if err != nil {
+				return nil, err
+			}
+			layers = append(layers, filepath.Join(blobsDir, strings.TrimPrefix(manifestDescriptor.Digest.String(), "sha256:")))
+			layers = append(layers, filepath.Join(blobsDir, strings.TrimPrefix(manifest.Config.Digest.String(), "sha256:")))
+			for _, layer := range manifest.Layers {
+				pathInPkg := filepath.Join(blobsDir, strings.TrimPrefix(layer.Digest.String(), "sha256:"))
+				layers = append(layers, pathInPkg)
+			}
+		}
+	}
+
+	return layers, nil
+}
```

Comment:
I think this function could be refactored to be shorter with some helpers (including potentially the one mentioned above)

Reply:
i did a little more than that